src/common/row_operations/row_matcher.cpp

```diff
@@ -0,0 +1,448 @@
+}
+
+template <>
+idx_t SelectComparison<LessThan>(Vector &left, Vector &right, const SelectionVector &sel, idx_t count,
+                                 SelectionVector *true_sel, SelectionVector *false_sel) {
+	return VectorOperations::DistinctLessThan(left, right, &sel, count, true_sel, false_sel);
+}
+
+template <>
+idx_t SelectComparison<LessThanEquals>(Vector &left, Vector &right, const SelectionVector &sel, idx_t count,
+                                       SelectionVector *true_sel, SelectionVector *false_sel) {
+	return VectorOperations::DistinctLessThanEquals(left, right, &sel, count, true_sel, false_sel);
+}
+
+template <bool NO_MATCH_SEL, class OP>
+static idx_t GenericNestedMatch(Vector &lhs_vector, const TupleDataVectorFormat &, SelectionVector &sel,
+                                const idx_t count, const TupleDataLayout &rhs_layout, Vector &rhs_row_locations,
+                                const idx_t col_idx, const vector<MatchFunction> &, SelectionVector *no_match_sel,
+                                idx_t &no_match_count) {
+	const auto &type = rhs_layout.GetTypes()[col_idx];
+
+	// Gather a dense Vector containing the column values being matched
+	Vector key(type);
+	const auto gather_function = TupleDataCollection::GetGatherFunction(type);
+	gather_function.function(rhs_layout, rhs_row_locations, col_idx, sel, count, key,
+	                         *FlatVector::IncrementalSelectionVector(), key, gather_function.child_functions);
+
+	// Densify the input column
+	Vector sliced(lhs_vector, sel, count);
+
+	if (NO_MATCH_SEL) {
+		SelectionVector no_match_sel_offset(no_match_sel->data() + no_match_count);
+		auto match_count = SelectComparison<OP>(sliced, key, sel, count, &sel, &no_match_sel_offset);
+		no_match_count += count - match_count;
+		return match_count;
+	} else {
+		return SelectComparison<OP>(sliced, key, sel, count, &sel, nullptr);
+	}
+}
+
+void RowMatcher::Initialize(const bool no_match_sel, const TupleDataLayout &layout, const Predicates &predicates) {
+	match_functions.reserve(predicates.size());
+	for (idx_t col_idx = 0; col_idx < predicates.size(); col_idx++) {
+		match_functions.push_back(GetMatchFunction(no_match_sel, layout.GetTypes()[col_idx], predicates[col_idx]));
+	}
+}
+
+idx_t RowMatcher::Match(DataChunk &lhs, const vector<TupleDataVectorFormat> &lhs_formats, SelectionVector &sel,
+                        idx_t count, const TupleDataLayout &rhs_layout, Vector &rhs_row_locations,
+                        SelectionVector *no_match_sel, idx_t &no_match_count) {
+	D_ASSERT(!match_functions.empty());
+	for (idx_t col_idx = 0; col_idx < match_functions.size(); col_idx++) {
+		const auto &match_function = match_functions[col_idx];
+		count =
+		    match_function.function(lhs.data[col_idx], lhs_formats[col_idx], sel, count, rhs_layout, rhs_row_locations,
+		                            col_idx, match_function.child_functions, no_match_sel, no_match_count);
```

Comment:
It took me a bit to understand `count` here and `sel`. Maybe we can add a comment explaining that `count` holds the current number of remaining rows that we have to compare and which are evaluated to `true` in all prior predicates. And that `sel` is a selection vector pointing to `count` rows that are still eligible?